.github/actions/build-sample-app/action.yml

```diff
@@ -0,0 +1,142 @@
+    required: false
+  jf-mvn-build-commands:
+    description: 'commands passed to \"jfrog rt mvn\" to publish the app'
+    required: false
+    default: '-B clean install -DskipTests'
+  docker-push:
+    description: 'whether or not to push docker image to docker hub'
+    type: boolean
+    required: false
+    default: true
+  docker-registry:
+    description: 'docker registry'
+    required: false
+  docker-username:
+    description: 'docker username'
+    required: false
+  docker-password:
+    description: 'docker password'
+    required: false
+  docker-images:
+    description: 'csv of docker images to push'
+    required: false
+  docker-images-override:
+    description: 'csv of docker image tags to use when pushing (parallel array to docker-images)'
+    required: false
+runs:
+  using: "composite"
+  steps:
+    - uses: actions/checkout@v2
+    - uses: jvalkeal/setup-maven@v1
+      with:
+        maven-version: 3.8.4
+    - uses: actions/setup-java@v2
+      with:
+        distribution: adopt
+        java-version: 8
+        cache: maven
+
+    - name: Maven build
+      shell: bash
+      run: ./mvnw ${{ inputs.mvn-build-commands }}
+      working-directory: ${{ inputs.app-dir }}
+
+    - name: Setup JFrog CLI
+      if: ${{ inputs.artifactory-publish == 'true' }}
+      uses: jfrog/setup-jfrog-cli@v1
+      with:
+        version: 1.46.4
+      env:
+        JF_ARTIFACTORY_SPRING: ${{ inputs.jf-artifactory-spring }}
+
+    - name: Configure JFrog Cli
+      if: ${{ inputs.artifactory-publish == 'true' }}
+      shell: bash
+      run: |
+        jfrog rt mvnc \
+        --server-id-resolve=repo.spring.io \
+        --server-id-deploy=repo.spring.io \
+        --repo-resolve-releases=libs-milestone \
+        --repo-resolve-snapshots=libs-snapshot \
+        --repo-deploy-releases=release \
+        --repo-deploy-snapshots=snapshot
+        echo JFROG_CLI_BUILD_NAME=spring-cloud-dataflow-samples >> $GITHUB_ENV
+        echo JFROG_CLI_BUILD_NUMBER=$GITHUB_RUN_NUMBER >> $GITHUB_ENV
+      working-directory: ${{ inputs.app-dir }}
+
+    - name: Publish to Artifactory
+      if: ${{ inputs.artifactory-publish == 'true' }}
+      shell: bash
+      run: |
+        jfrog rt mvn ${{ inputs.jf-mvn-build-commands }}
+        jfrog rt build-publish
+      working-directory: ${{ inputs.app-dir }}
+
+    - name: Login Dockerhub
+      if: ${{ inputs.docker-push == 'true' }}
+      uses: docker/login-action@v1
+      with:
+        registry: ${{ inputs.docker-registry }}
+        username: ${{ inputs.docker-username }}
+        password: ${{ inputs.docker-password }}
+
+    - name: Push images to Dockerhub
+      if: ${{ inputs.docker-push == 'true' }}
+      env:
+        DOCKER_REGISTRY: ${{ inputs.docker-registry }}
+        DOCKER_IMAGES: ${{ inputs.docker-images }}
+        DOCKER_IMAGES_OVERRIDE: ${{ inputs.docker-images-override }}
+      shell: bash
+      run: |
+        # read the images into arrays
```

Comment:
This is just a little too much bash scripting that is not automatically tested - it makes me nervous. I did test it rigorously locally on all of the use cases though.

Reply:
I may move this to a script file in the actions dir later.